parsec/basicclient.go

```diff
@@ -15,23 +16,68 @@ import (
 // BasicClient is a Parsec client representing a connection and set of API implementations
 type BasicClient struct {
 	opclient         *operations.Client
-	auth             auth.AuthenticationType
+	auth             Authenticator
 	implicitProvider ProviderID
+	config           *ClientConfig
 }
 
 // InitClient initializes a Parsec client
-func InitClient() (*BasicClient, error) {
-	opclient, err := operations.InitClient()
+// This will autoselect the first provider returned by the parsec service.  It will also attempt to
+// select the first available authenticator it can configure.  The config can either be a *ClientConfig or a string.
+// If it is a string, then this is used as an application name for a default Direct Authenticator.
+// If nil is passed, then the client will try and find the first supported authenticator that requires no configuration.
+func InitClient(config interface{}) (*BasicClient, error) {
```

Comment:
Do you think having a configuration is useful at this stage? I am wondering if in most use-cases ppl will just want to use the default. Maybe change the endpoint path through the `PARSEC_SERVICE_ENDPOINT` variable but that's supported.

I like the tradeoff that we have in the Rust client:
- either you can create a "naked" client where core provider + no authentication are set
- either you create a default client (with a string as parameter) where everything is chosen automatically

and then you still have methods to set a new connection/authentication/implicit provider.

So that it makes the interface a bit simpler and remove the need for the configuration code? What do you think?

Reply:
ahh, so that's what the naked client is?

Can add CreateNakedClient, CreateDefaultClient (which is current InitClient with a string passed).

I quite like the the configuration - tend to find it more natural than a load of mutator calls.  Maybe CreateNakedClient and CreateConfiguredClient ?

Reply:
Have done the latter- CreateNakedClient and CreateConfiguredClient - gives same functionality as rust client, and adds config, in case people like that too.

Reply:
> ahh, so that's what the naked client is?

indeed 😬 

> Have done the latter- CreateNakedClient and CreateConfiguredClient - gives same functionality as rust client, and adds config, in case people like that too.

Thanks yeah. I really like having the option there 👍 !